python/src/aiconfig/callback.py

```diff
@@ -0,0 +1,139 @@
+# Standard Libraries
+import asyncio
+import json
+import time
+
+# Third Party Libraries
+from result import Err, Ok
+from pydantic import BaseModel, ConfigDict
+from typing import (
+    Coroutine,
+    Final,
+    List,
+    Sequence,
+    TextIO,
+    TypeAlias,
+    Union,
+    Any,
+    Awaitable,
+    Callable,
+    Optional,
+)
+
+# Constants
+DEFAULT_TIMEOUT = 5  # Default timeout for callback execution in seconds
+
+
+class CallbackEvent:
+    """
+    Represents an event with data to be passed to a callback.
+    """
+
+    def __init__(self, name: str, data: Any, ts_ns: int = time.time_ns()):
+        self.name = name
+        # Anything available at the time the event happens.
+        # It is passed to the callback.
+        self.data = data
+        self.ts_ns = ts_ns
+
+
+# Type Aliases
+Callback = Callable[[CallbackEvent], Awaitable]
+Result: TypeAlias = Union[Ok[Any], Err[Any]]
+
+
+async def execute_coroutine_with_timeout(
+    coroutine: Coroutine[Any, Any, Any], timeout: int
+) -> Result:
+    """
+    Execute a coroutine with a timeout, return an Ok result or an Err on Exception
+
+    Args:
+        coroutine: The coroutine to execute
+        timeout: The timeout in seconds
+    
+    Returns:
+        An Ok result if the coroutine executes successfully or an Err on Exception
+    """
+    try:
+        task = asyncio.create_task(coroutine)
+        res = await asyncio.wait_for(task, timeout=timeout)
+        return Ok(res)
+    except Exception as e:
+        return Err(str(e))
+
+
+class Record(BaseModel):
+    model_config = ConfigDict(strict=True, frozen=True)
+
+
+class pydanticBackedCallBackEvent(Record):
+    """
+    Data class that represents an event that triggers a callback. A copy of CallbackEvent
+    but backed by a pydantic model for easy serialization.
+    """
+    name: str
+    data: Any
+    ts_ns: int = time.time_ns()
+
+
+class CallbackManager:
+    """
+    Manages a sequence of callbacks to be executed in response to Events
+    """
+    def __init__(self, callbacks: Sequence[Callback], timeout: int = DEFAULT_TIMEOUT) -> None:
```

Comment:
One thing to consider about default arguments: if the user accidentally passes down None, it will overwrite this. None. is often used as a sentinel for "use whatever the default is" so that would result in unexpected behavior.

I usually set to the real value inside the function for this reason.